mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -117,59 +99,32 @@ public suspend fun runMosaic(body: suspend MosaicScope.() -> Unit): Unit = corou
 					size = IntSize(terminal.info.width, terminal.info.height),
 				)
 			}
-			delay(50)
+			delay(50L)
 		}
 	}
 
-	coroutineScope {
-		val scope = object : MosaicScope, CoroutineScope by this {
-			override fun setContent(content: @Composable () -> Unit) {
-				composition.setContent {
-					CompositionLocalProvider(LocalTerminal provides terminalInfo.value) {
-						content()
-					}
-				}
-			}
-		}
+	val rootNode = createRootNode()
+	val applier = MosaicNodeApplier(rootNode) {
+		val render = rendering.render(rootNode)
+		platformDisplay(render)
+	}
 
-		var snapshotNotificationsPending = false
-		val observer: (state: Any) -> Unit = {
-			if (!snapshotNotificationsPending) {
-				snapshotNotificationsPending = true
-				launch {
-					snapshotNotificationsPending = false
-					Snapshot.sendApplyNotifications()
-				}
-			}
-		}
-		val snapshotObserverHandle = Snapshot.registerGlobalWriteObserver(observer)
-		try {
-			scope.body()
-		} finally {
-			snapshotObserverHandle.dispose()
+	Composition(applier, recomposer).setContent {
+		CompositionLocalProvider(LocalTerminal provides terminalInfo.value) {
+			content()
 		}
 	}
 
-	// Ensure the final state modification is discovered. We need to ensure that the coroutine
-	// which is running the recomposition loop wakes up, notices the changes, and waits for the
-	// next frame. If you are using snapshots this only requires a single yield. If you are not
-	// then it requires two yields. THIS IS NOT GREAT! But at least it's implementation detail...
-	// TODO https://issuetracker.google.com/issues/169425431
-	yield()
-	yield()
-	Snapshot.sendApplyNotifications()
-	yield()
-	yield()
-
-	if (hasFrameWaiters) {
-		CompletableDeferred<Unit>().also {
-			displaySignal = it
-			it.await()
-		}
+	val effectJob = checkNotNull(recomposer.effectCoroutineContext[Job]) {
+		"No Job in effectCoroutineContext of recomposer"
 	}
+	effectJob.children.forEach { it.join() }
```

Comment:
Compose UI for Desktop does not appear to be doing this for its `application { }` block and `Window`s. Is this not included in the `close()`/`join()` dance below? Granted, they are using subcompositions, so perhaps they just don't anticipate anyone launching effects in the root and have a bug?

Reply:
I looked more closely at the implementations and in general it was possible to preserve the behavior without these instructions, so I removed them.
I don’t remember for sure why they appeared in the first place, but I don’t rule out that updates could have had an impact.